### ISSUE TITLE
Invalid opcodes in rectavoider.svc16

### ISSUE THREAD

**Issue by xyloflake** [JanNeuendorf/SVC16#4]
Hello, cool repo you've got here :)

I have created a simple disassembler that just reads bytes from the svc16 file as an `Vec<u8>` buffer and then converts them to an `Vec<u16>` containing the instructions while using `from_le_bytes`.

I then iterate through the u16 vec taking 4 values at a time. Then map through them and I have an array [String; 0x10] called op_codes that basically has the opcodes listed in the readme file in order. The disassembler then matches the 1st out of the 4 values to the op_codes and replaces them with the string version of the opcode (for example 0 gets replaced with "Set").

The disassembler works fine with `all_colors.svc16` but **throws invalid opcode when disassembling `rectavoider.svc16`**. Upon further inspection, I found out that **the file has invalid opcodes, the first occurence of which occurs at line 4711** after executing
```bash
hexyl rectavoider.svc16 -pv --panels 1 > hex.txt
```
The result is:
![Screenshot_20241219-132803-1.jpg](https://github.com/user-attachments/assets/20cf8772-40ca-4a05-8eb8-e3d2a7708c95)

The opcode observed is `c8 00` in hex which translates to `200` in decimal. From my understanding, there is no opcode beyond 16(?).

On further inspection, dozens of such invalid opcodes were observed.

Is there anything I'm missing out on or doing wrong?

I obviously have executed `rectavoider.svc16` with svc16 and the game works as expected.

Any help is much appreciated and thanks for creating this amazing repo!

**Comment by JanNeuendorf** [comment]
Hello,
first of all **cool work**!
What is happening here is that the game does not just contain instructions.
It also contains the data for the sprites. This is fine, as long as I never try to run them as code. 
I stored the picture of the galaxy in the background by using the individual bits of the values so I get 16 pixels per value. This is why the values are bigger than 15. There are also some other ""literal values" already written in. 
I could not reproduce which one you would first encounter at 4711 (,but the range sounds about right).

(One thing that I  should add is that the demo-game is 1. in no way compiled in an efficient or even sensible way and 2. I think it might contain unused code. )

**Comment by xyloflake** [comment]
> first of all **cool work**!

Thank you so much Jan! Do you mind if I publish it to github once it's in a usable state? However I think people should build that themselves but doesn't hurt to put it out there(?).

> What is happening here is that the game does not just contain instructions.
> It also contains the data for the sprites. This is fine, as long as I never try to run them as code. 
> I stored the picture of the galaxy in the background by using the individual bits of the values so I get 16 pixels per value. This is why the values are bigger than 15. There are also some other ""literal values" already written in. 

Ah cool, I expected this! So in order to improve my disassembler, I basically have to track what code never runs? If I do manage to track it down like that, how do I store these values such that when it's assembled back, it runs fine? That would need the use of something like section.data in x64 asm or in simpler terms, variables?

What I am thinking about is, there's no official assembly for svc16 right. Like there could be a very close one that looks like assembly and has basic stuff like variables and stuff baked in?

> (One thing that I  should add is that the demo-game is 1. in no way compiled in an efficient or even sensible way and 2. I think it might contain unused code. )

Oh yea I completely am aware of that, thanks! I am quite interested in how you got the text rendering working. Is it something like you store bitmaps or images of numbers and text and then just display them or have you got a rasterizer 👀?

Also, are you supposed to make a compiler for an existing lang like C or make an entirely new lang because of the marginal differences?

I was trying to disassemble this to get an idea of how the text rendering is done (like the score, "game over", etc) 😅

**Comment by JanNeuendorf** [comment]
> Do you mind if I publish it to github once it's in a usable state? However I think people should build that themselves but doesn't hurt to put it out there(?).

Yes, please put it out there! I think that having an inspector for a rom would be completely universal. 
I think that this could also be in this repo either as a separate program or a subcommand of the emulator. But that is up to you.  

> Ah cool, I expected this! So in order to improve my disassembler, I basically have to track what code never runs? If I do manage to track it down like that, how do I store these values such that when it's assembled back, it runs fine? That would need the use of something like section.data in x64 asm or in simpler terms, variables?

I think, unfortunately, there is no way to determine if something is supposed to be read as instructions or data from just looking at the rom. The program could compute an arbitrary address at runtime and then set the instruction pointer to that value. **This also means that the address of the opcodes does not have to be divisible by four!** 
I am sure there is a good way to present it in a helpful way anyway. 

> What I am thinking about is, there's no official assembly for svc16 right. Like there could be a very close one that looks like assembly and has basic stuff like variables and stuff baked in?

I have made something like this for the test programs. I can explain it if you are interested but I am not married to it. It is just a convention after all. I am pretty open to the idea of making a good assembler available. I just said that I am not providing one in the Readme to make it clear that the specification is completely neutral. We could make a different directory in this repo for the assembler and make it clear that it is only one possible implementation? (Or a different repo for developing tooling.)

> Oh yea I completely am aware of that, thanks! I am quite interested in how you got the text rendering working. Is it something like you store bitmaps or images of numbers and text and then just display them or have you got a rasterizer 👀?

No, it is really just bitmaps of characters. Hopefully at some time in the future there could be a working rasterizer. 
(The "Game Over" is just one sprite.)

> Also, are you supposed to make a compiler for an existing lang like C or make an entirely new lang because of the marginal differences?

Whatever seems more interesting to you.

**Comment by xyloflake** [comment]
> Yes, please put it out there! I think that having an inspector for a rom would be completely universal. 
> I think that this could also be in this repo either as a separate program or a subcommand of the emulator. But that is up to you.  

Thanks! I will think about it and let you know. I am currently leaning towards having a different repo that has tools for development of svc16 and this repo contains the specification and the main emulator (however I think both should be in different repos for obvious reasons). I would suggest creating an organisation and putting all of the repos in one place 😄.

> I think, unfortunately, there is no way to determine if something is supposed to be read as instructions or data from just looking at the rom. The program could compute an arbitrary address at runtime and then set the instruction pointer to that value. **This also means that the address of the opcodes does not have to be divisible by four!** 
> I am sure there is a good way to present it in a helpful way anyway. 

Makes sense. Maybe the specification can be altered to NOT have arbitrary ways to store data. Like the specification could have something like a much more solid concept towards storing data in the svc16 representation. The current specfication leads to unexpected behaviour as you described and this would really violate simplicity. It would be better if there's something like a separate storage buffer for storing stuff like this? Maybe there could be a better way, I don't really know. Exploring the possibility should be a feasible option.

> I have made something like this for the test programs. I can explain it if you are interested but I am not married to it. It is just a convention after all. I am pretty open to the idea of making a good assembler available. I just said that I am not providing one in the Readme to make it clear that the specification is completely neutral. We could make a different directory in this repo for the assembler and make it clear that it is only one possible implementation? (Or a different repo for developing tooling.)

Please explain! I would be much obliged to have an explanation and if you'd like, the code that compiles rectavoider to have an idea of how you're doing all this!

Like I said, a different repo for all the tooling stuff baked in would be much better from my POV ^_^

> No, it is really just bitmaps of characters. Hopefully at some time in the future there could be a working rasterizer. 
> (The "Game Over" is just one sprite.)

Ah, I have to say that's quite disappointing (I thought you already got text rendering working haha XD). Jokes aside, I am not really sure if this current specification would hold up to rasterization. Storing the fonts in the binary itself is a very bad idea and I would suggest implementing file system stuff to make it possible to load stuff from the file system. And this would be much better as you stop having data in the machine code instead of just instructions. Obviously this introduces some complexity but I think it's a necessary step to making complex stuff.

> Whatever seems more interesting to you. 

I was kind of asking if it's even possible to implement all features implemented in C?

Thanks for the replies, I appreciate it ^_^

**Comment by JanNeuendorf** [comment]
> Thanks! I will think about it and let you know. I am currently leaning towards having a different repo that has tools for development of svc16 and this repo contains the specification and the main emulator (however I think both should be in different repos for obvious reasons). I would suggest creating an organisation and putting all of the repos in one place 😄.

I created a repo (We can figure that organisation thing later). 

> Makes sense. Maybe the specification can be altered to NOT have arbitrary ways to store data. Like the specification could have something like a much more solid concept towards storing data in the svc16 representation. The current specfication leads to unexpected behaviour as you described and this would really violate simplicity. It would be better if there's something like a separate storage buffer for storing stuff like this? Maybe there could be a better way, I don't really know. Exploring the possibility should be a feasible option.

To be honest, I am not open to changing the architecture. It is supposed to be a bit weird. The idea is that the programmer has this block of memory and they have to figure out what to do with it. Where should the functions go? Where should the data go? Does the data need to be compressed to fit into the available memory? This is all up to them to figure out.

>Please explain! I would be much obliged to have an explanation and if you'd like, the code that compiles rectavoider to have an idea of how you're doing all this!

My "assembly" looks a bit like this:
```
# main
set _color 200
cmp "0" _mouse_key 502
skip 2 0 502
	set _color 5000
cmp "1" _mouse_key 502
skip 2 0 502
	set _color 50000
set _counter 0 
	cmp "65534" _counter _loop_done
	print _color _counter
	add "1" _counter _counter
skip 0 3 _loop_done

print "65534" _mouse_pos
set _x_mask 255 
band _x_mask _mouse_pos _mouse_x
sub _mouse_pos _mouse_x _mouse_y
div _mouse_y "256" _mouse_y
set _counter 0 
	mul _mouse_y "256" _current_cross
	add _current_cross _counter _current_cross
	print "65534" _current_cross
	add _counter "1" _counter
	cmp _x_mask _counter _loop_done
skip 0 5 _loop_done

set _counter 0 
	mul _counter "256" _current_cross
	add _current_cross _mouse_x _current_cross
	print "65534" _current_cross
	add _counter "1" _counter
	cmp _x_mask _counter _loop_done
skip 0 5 _loop_done

sync _mouse_pos _mouse_key
goto 0

```
(This is a program I used to test mouse input). Then the program does a couple of things to run this:
- replace the opcodes with numbers
- find the "literal numbers" and put their value somewhere into the rom. They are then replaced with that address. 
- The position of the `#blocks` is also available as literals, that makes it possible to call functions, because they can be written as separate blocks. 

That is the basic idea. 

>Ah, I have to say that's quite disappointing (I thought you already got text rendering working haha XD). Jokes aside, I am not really sure if this current specification would hold up to rasterization. Storing the fonts in the binary itself is a very bad idea and I would suggest implementing file system stuff to make it possible to load stuff from the file system. And this would be much better as you stop having data in the machine code instead of just instructions. Obviously this introduces some complexity but I think it's a necessary step to making complex stuff.

Storing the fonts in the ROM is **exactly** what one would have to do. All information to emulate the program must be contained in the ROM itself. So it runs the same everywhere. The last sentence I want to say is "It works on my virtual machine". So filesystem access is definitely out of the question. I think that font rasterization might actually be possible.

>I was kind of asking if it's even possible to implement all features implemented in C? 
you don't get floating point operations or signed operations natively but you could fake it in some ways. 
I believe that you could at least have a language that looks a lot like C at some point, but I have not built anything that complex yet.

**Comment by xyloflake** [comment]
> I created a repo (We can figure that organisation thing later). 

Great move, thanks for the invite!

> To be honest, I am not open to changing the architecture. It is supposed to be a bit weird. The idea is that the programmer has this block of memory and they have to figure out what to do with it. Where should the functions go? Where should the data go? Does the data need to be compressed to fit into the available memory? This is all up to them to figure out.

Ah okay, I get it now. I had a vague view point of the goals this project had but now I have a clear view of the goals. It should be simple but you're supposed to do the critical thinking involved yourself so there's no "magic" happening.
 
> My "assembly" looks a bit like this:
> ```
> # main
> set _color 200
> cmp "0" _mouse_key 502
> skip 2 0 502
> 	set _color 5000
> cmp "1" _mouse_key 502
> skip 2 0 502
> 	set _color 50000
> set _counter 0 
> 	cmp "65534" _counter _loop_done
> 	print _color _counter
> 	add "1" _counter _counter
> skip 0 3 _loop_done
> 
> print "65534" _mouse_pos
> set _x_mask 255 
> band _x_mask _mouse_pos _mouse_x
> sub _mouse_pos _mouse_x _mouse_y
> div _mouse_y "256" _mouse_y
> set _counter 0 
> 	mul _mouse_y "256" _current_cross
> 	add _current_cross _counter _current_cross
> 	print "65534" _current_cross
> 	add _counter "1" _counter
> 	cmp _x_mask _counter _loop_done
> skip 0 5 _loop_done
> 
> set _counter 0 
> 	mul _counter "256" _current_cross
> 	add _current_cross _mouse_x _current_cross
> 	print "65534" _current_cross
> 	add _counter "1" _counter
> 	cmp _x_mask _counter _loop_done
> skip 0 5 _loop_done
> 
> sync _mouse_pos _mouse_key
> goto 0
> 
> ```
> (This is a program I used to test mouse input). Then the program does a couple of things to run this:
> - replace the opcodes with numbers
> - find the "literal numbers" and put their value somewhere into the rom. They are then replaced with that address. 
> - The position of the `#blocks` is also available as literals, that makes it possible to call functions, because they can be written as separate blocks. 
> 
> That is the basic idea. 

Amazing to see the code! This however makes disassembly very hard since there's no specific rule to how the programs assembling the "assembly" work. So the assembler that allocates memory based on how the assembler wants to allocate memory. So one program compiled by one assembler may not be disassemblable by a "general" disassembler? In that case we should first make a specification for an assembly type intermediate language and then point out to examples as "this code was assembled using svct's assembler and can be disassembled using svct's specific disassemblers only". That should make sense as real computers have a very strict assembly syntax and obviously if you create an assembler that works differently, you may not be able to disassemble it.

For example x86 disassemblers can't disassemble x64 machine code because standards of both are different and the underlying machine code would be different.

Implementing a rom analyzer should be feasible but I think it should come after the assembler and disassembler?

> Storing the fonts in the ROM is **exactly** what one would have to do. All information to emulate the program must be contained in the ROM itself. So it runs the same everywhere. The last sentence I want to say is "It works on my virtual machine". So filesystem access is definitely out of the question. I think that font rasterization might actually be possible.

I get it. What I meant by font rasterization being not possible is, you'd have a very limited set of characters with a 131kb file limit. Some fonts are well above the limit and obviously they could be compressed but then you don't have space for other stuff you'd like to store. However as you've stated, this should be all up to the programmer and I undeniably agree. It should be actually possible to make a **small** rasterizer.

> you don't get floating point operations or signed operations natively but you could fake it in some ways. 
> I believe that you could at least have a language that looks a lot like C at some point, but I have not built anything that complex yet. 

Yea I get the faking "flop"s stuff. It will look like C but it can't be exactly C I think as it's pretty much limited. The architecture doesn't have SIMD and other native stuff for obvious reasons. I propose we make it close to C but not exactly C.

**Comment by JanNeuendorf** [comment]
>Amazing to see the code! This however makes disassembly very hard since there's no specific rule to how the programs assembling the "assembly" work. So the assembler that allocates memory based on how the assembler wants to allocate memory. So one program compiled by one assembler may not be disassemblable by a "general" disassembler? In that case we should first make a specification for an assembly type intermediate language and then point out to examples as "this code was assembled using svct's assembler and can be disassembled using svct's specific disassemblers only".

Yes, this sounds perfectly reasonable.

>Implementing a rom analyzer should be feasible but I think it should come after the assembler and disassembler?

Yes, I do not really know what a general rom-analyzer would really do that you could not do with *hexdump*. 

>I get it. What I meant by font rasterization being not possible is, you'd have a very limited set of characters with a 131kb file limit. Some fonts are well above the limit and obviously they could be compressed but then you don't have space for other stuff you'd like to store. However as you've stated, this should be all up to the programmer and I undeniably agree. It should be actually possible to make a small rasterizer.

Without having any knowledge to base my opinion on, I would guess that this would not be a problem of memory. The problem is that with the 3000000 instruction limit, you only get about 45 instructions per pixel. This might not be enough for complicated rasterization. It is, however, enough to index a bitmap. So the much easier way is to include bitmaps for characters.

> Yea I get the faking "flop"s stuff. It will look like C but it can't be exactly C I think as it's pretty much limited. The architecture doesn't have SIMD and other native stuff for obvious reasons. I propose we make it close to C but not exactly C.

I think this is a good idea if we want to have a good example project. It would show how you can implement something very *normal*. 

One question I have here, however, is the order of operations. I think it is very important to lock in a final version of the specification for the system itself very early. I guess that no one (including you) would want to invest many hours into building something for a system that could still change. I think that there is nothing standing in the way of this except for finishing the pdf document.

**Comment by xyloflake** [comment]
> Yes, this sounds perfectly reasonable.

Cool, we will stick to this then.

> Yes, I do not really know what a general rom-analyzer would really do that you could not do with *hexdump*. 

I'm not sure either. I don't know what rom analyzers do.

> Without having any knowledge to base my opinion on, I would guess that this would not be a problem of memory. The problem is that with the 3000000 instruction limit, you only get about 45 instructions per pixel. This might not be enough for complicated rasterization. It is, however, enough to index a bitmap. So the much easier way is to include bitmaps for characters.

I have created couple of renderers over the past year and I can say that 45 instructions per pixels *might* be enough. I don't think text falls into complex rasterization, and that is why it shouldn't matter. Why I say it would be a problem of memory is the involvement of complex stuff like graphemes and glyphs. There would be many things to store in memory and I am afraid we might hit OOM. But this is only theory and theory can only take us so far haha. **We should worry about this after having the compiler to maintain order of priorities.**

> I think this is a good idea if we want to have a good example project. It would show how you can implement something very *normal*. 

I absolutely agree, a good example project is definitely needed to help people figure out what they can build.

> One question I have here, however, is the order of operations. I think it is very important to lock in a final version of the specification for the system itself very early. 

Fantastic thought. I agree that the specification should be deemed complete and no major changes being introduced after being declared stable.

> I guess that no one (including you) would want to invest many hours into building something for a system that could still change. I think that there is nothing standing in the way of this except for finishing the pdf document. 

Well, as for me, I think I can adapt but a systematic approach would be to complete the specfication first.

Thanks for all the good stuff!

**Comment by JanNeuendorf** [comment]
Great!
I opened an issue about the question of adding sound, which would be my last blocker. If this is solved, the specifications are only a matter of expanding the writing I already copied from the readme. If you want to help, you have probably found the *.typ* document. (If you have not heard of it, it can be compiled with **typst**).
Then thank you very much for collaborating on this and see you in another thread!

**Comment by xyloflake** [comment]
> I opened an issue about the question of adding sound, which would be my last blocker. If this is solved, the specifications are only a matter of expanding the writing I already copied from the readme. If you want to help, you have probably found the *.typ* document. (If you have not heard of it, it can be compiled with **typst**).

@JanNeuendorf would love to help but I'm not sure how I should write the specification. What more is there to be added? Yes, I have found the .typ document and have taken a look at it and I've used typst before so nothing to worry about.
